pkg/storageops/vsphere/vsphere.go

```diff
@@ -0,0 +1,376 @@
+	if err != nil {
+		logrus.Errorf("failed to get device path for disk: %s on vm: %s", diskPath, vmObj.Name())
+		return "", err
+	}
+
+	return path.Join(diskByIDPath, diskSCSIPrefix+diskUUID), nil
+}
+
+func (ops *vsphereOps) Enumerate(volumeIds []*string,
+	labels map[string]string,
+	setIdentifier string,
+) (map[string][]interface{}, error) {
+	return nil, storageops.ErrNotSupported
+}
+
+// Snapshot the volume with given volumeID
+func (ops *vsphereOps) Snapshot(volumeID string, readonly bool) (interface{}, error) {
+	return nil, storageops.ErrNotSupported
+}
+
+// SnapshotDelete deletes the snapshot with given ID
+func (ops *vsphereOps) SnapshotDelete(snapID string) error {
+	return storageops.ErrNotSupported
+}
+
+// ApplyTags will apply given labels/tags on the given volume
+func (ops *vsphereOps) ApplyTags(volumeID string, labels map[string]string) error {
+	return storageops.ErrNotSupported
+}
+
+// RemoveTags removes labels/tags from the given volume
+func (ops *vsphereOps) RemoveTags(volumeID string, labels map[string]string) error {
+	return storageops.ErrNotSupported
+}
+
+// Tags will list the existing labels/tags on the given volume
+func (ops *vsphereOps) Tags(volumeID string) (map[string]string, error) {
+	return nil, storageops.ErrNotSupported
+}
+
+// GetVMObject fetches the VirtualMachine object corresponding to the given virtual machine uuid
+func GetVMObject(ctx context.Context, conn *vclib.VSphereConnection, vmUUID string) (*vclib.VirtualMachine, error) {
+	// TODO change impl below using multiple goroutines and sync.WaitGroup to make it faster
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	if err := conn.Connect(ctx); err != nil {
+		return nil, err
+	}
+
+	if len(vmUUID) == 0 {
+		return nil, fmt.Errorf("virtual machine uuid is required")
+	}
+
+	datacenterObjs, err := vclib.GetAllDatacenter(ctx, conn)
+	if err != nil {
+		return nil, err
+	}
+
+	// Lookup in each vsphere datacenter for this virtual machine
+	for _, dc := range datacenterObjs {
+		vm, err := dc.GetVMByUUID(ctx, vmUUID)
+		if err != nil {
+			if err != vclib.ErrNoVMFound {
+				logrus.Warnf("failed to find vm with uuid: %s in datacenter: %s due to err: %v", vmUUID, dc.Name(), err)
+				// don't let one bad egg fail entire search. keep looking.
+			} else {
+				logrus.Debugf("did not find vm with uuid: %s in datacenter: %s", vmUUID, dc.Name())
+			}
+			continue
+		}
+
+		if vm != nil {
+			host, err := vm.HostSystem(ctx)
+			if err != nil {
+				return nil, err
+			}
+
+			logrus.Infof("vm: %s uuid: %s is in datacenter: %s running on host: %v", vm.Name(), vmUUID, dc.Name(), host.Reference())
+			return vm, nil
+		}
+	}
+
+	return nil, fmt.Errorf("failed to find vm with uuid: %s in any datacenter for vc: %s", vmUUID, conn.Hostname)
+}
+
+func (ops *vsphereOps) renewVM(ctx context.Context, vm *vclib.VirtualMachine) (*vclib.VirtualMachine, error) {
+	err := ops.conn.Connect(ctx)
```

Comment:
Connect implementation refreshes the Client object or creates a new one. If it does not return an error, then we can pass in ops.conn.GoVmomiClient to RenewVM.

If it returns an error then we can call the NewClient. Otherwise we are unnecessarily creating new clients for every renew call.

Reply:
good point. Fixed.